ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/main.rs

```diff
@@ -15,11 +34,174 @@ pub struct ParserConfig {
+) -> anyhow::Result<()> {
+    todo!();
+}
+
+/**
+ * Spawns a worker to pull from Channel and perform parsing operations
+ */
+async fn _spawn_parser(
+    _id: usize,
+    _semaphore: Arc<Semaphore>,
+    _receiver: Arc<Mutex<Receiver<(NFTMetadataCrawlerEntry, String)>>>,
+    _conn: Pool<ConnectionManager<PgConnection>>,
+    _bucket: String,
+    mut _grpc_client: SubscriberClient<Channel>,
+    _subscription_name: String,
+    _cdn_prefix: String,
+) -> anyhow::Result<()> {
+    todo!();
+}
+
+/**
+ * Subscribes to PubSub and sends URIs to Channel
+ */
+async fn process_response(
+    sender: Sender<(NFTMetadataCrawlerEntry, String)>,
+    mut grpc_client: SubscriberClient<Channel>,
+    subscription_name: String,
+) -> anyhow::Result<()> {
+    let ts = get_token_source().await;
+    loop {
+        // Pulls an entry from PubSub
+        let entries = consume_uris(
+            10,
+            &mut grpc_client,
+            subscription_name.clone(),
+            ts.token().await?.access_token,
+        )
+        .await?;
+
+        // Parses entry and sends to Channel
+        for msg in entries.into_inner().received_messages {
+            if let Some(entry_option) = msg.message {
+                let entry = entry_option.data;
+                let ack = msg.ack_id;
+                let entry = NFTMetadataCrawlerEntry::new(String::from_utf8(entry)?)?;
+                sender.send((entry.clone(), ack))?;
+            }
+        }
+    }
+}
+
+/**
+ * Spawns a worker to pull from Channel and perform parsing operations
+ */
+async fn spawn_parser(
+    id: usize,
+    semaphore: Arc<Semaphore>,
+    receiver: Arc<Mutex<Receiver<(NFTMetadataCrawlerEntry, String)>>>,
+    conn: Pool<ConnectionManager<PgConnection>>,
+    bucket: String,
+    mut grpc_client: SubscriberClient<Channel>,
+    subscription_name: String,
+    cdn_prefix: String,
+) -> anyhow::Result<()> {
+    let ts = get_token_source().await;
+    loop {
+        let _ = semaphore.acquire().await?;
+
+        // Pulls entry from Channel
+        let (entry, ack) = receiver.lock().await.recv()?;
+        let token = ts.token().await?.access_token;
+
+        // Parses entry
+        info!(worker_id = id, "Received entry");
+        let mut parser = Parser::new(
+            entry,
+            bucket.clone(),
+            token.clone(),
+            conn.get()?,
+            cdn_prefix.clone(),
+        );
+        parser.parse().await?;
+
+        // Sends ack to PubSub
+        info!(worker_id = id, "Finished parsing");
+        send_acks(
+            vec![ack],
+            &mut grpc_client,
+            subscription_name.clone(),
+            token,
+        )
+        .await?;
+
+        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+    }
+}
+
+/**
+ * Main driver function
+ */
 #[async_trait::async_trait]
 impl RunnableConfig for ParserConfig {
     /// Main driver function
     async fn run(&self) -> anyhow::Result<()> {
```

Comment:
It would help to also summarize the function in the comment even if you have comments within the function. Ideally I don't have to read the code to find out what the function does.